File: 216. Combination Sum III/216.py
"""
Find all possible combinations of k numbers that add up to a number n,
given that only numbers from 1 to 9 can be used and each combination should be a unique set of numbers.

Ensure that numbers within the set are sorted in ascending order.
"""
class Solution:
    # @param {integer} k
    # @param {integer} n
    # @return {integer[][]}
    def combinationSum3(self, k, n):
        self.result = []
        self.path = []
        self.dfs(k, n, 1)
        return self.result

    def dfs(self, k, gap, start):
        if len(self.path) == k:
            if gap == 0:
                self.result.append(self.path[:])
            return
        for i in range(start,10):
            if gap < i:
                return
            self.path.append(i)
            self.dfs(k, gap - i, i + 1)
            self.path.pop()

s = Solution()
print(s.combinationSum3(3,7))
print(s.combinationSum3(3,3))

File: 216. Combination Sum III/216.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> result;
        vector<int> path;
        dfs(k, n, 1, path, result);
        return result;
    }

    void dfs(int k, int n, int cur, vector<int>& path, vector<vector<int>> &result) {
        if (path.size() == k) {
            if (n == 0) {
                result.emplace_back(path);
            }
            return;
        }

        for (int i = cur; i < 10; i++) {
            if (i > n) {
                return;
            }
            path.emplace_back(i);
            dfs(k, n - i, i + 1, path, result);
            path.pop_back();
        }
    }
};
